### PR TITLE
ROX-9398

### PR DESCRIPTION
**Pull request by daynewlee**
## Description

A detailed explanation of the changes in your PR.

Feel free to remove this section if it is overkill for your PR, and the title of your PR is sufficiently descriptive.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

TODO(replace-me)
Use this space to explain how you tested your PR, or, if you didn't test it, why
you did not do so. Valid reasons include, for example, "CI is sufficient",
"No testable changes". Feel free to attach JSON snippets, curl commands,
screenshots.

In addition to reviewing your code, reviewers **must** also review your testing
instructions and make sure they are sufficient.

### CODE REVIEW

File: central/imageintegration/store/postgres/gen.go [hunk 1]
```diff
@@ -1,3 +1,3 @@
 package postgres
 
-//go:generate pg-table-bindings-wrapper --type=storage.ImageIntegration --get-all-func
+//go:generate pg-table-bindings-wrapper --type=storage.ImageIntegration --search-category IMAGE_INTEGRATIONS --references storage.ImageIntegration --get-all-func
```
Comment (md2119) [inline central/imageintegration/store/postgres/gen.go:3]: This is a self-reference

File: central/cluster/datastore/datastore_impl.go [hunk 2]
```diff
@@ -152,6 +155,21 @@ func (ds *datastoreImpl) UpdateClusterStatus(ctx context.Context, id string, sta
 	return ds.clusterStorage.Upsert(ctx, cluster)
 }
 
+func (ds *datastoreImpl) removeImageIntegrationByClusterId(ctx context.Context, cluster *storage.Cluster) {
+
+	q := pkgSearch.NewQueryBuilder().AddExactMatches(pkgSearch.ClusterID, cluster.GetId()).ProtoQuery()
+	imageintegrations, err := ds.imageintegrationDataStore.Search(ctx, q)
+	if err != nil {
+		log.Errorf("failed to get image integrations for removed cluster %s: %v", cluster.GetId(), err)
```
Comment (md2119) [inline central/cluster/datastore/datastore_impl.go:163]: return

File: proto/api/v1/search_service.proto [hunk 3]
```diff
@@ -56,6 +56,7 @@ enum SearchCategory {
     COMPLIANCE_DOMAIN         = 44;
     CLUSTER_HEALTH            = 45;
     POLICY_CATEGORIES         = 46;
+    IMAGE_INTEGRATIONS         = 47;
```
Comment (RTann) [inline proto/api/v1/search_service.proto:59]: nit: align the `= 47;` with the rest

File: central/imageintegration/datastore/internal/search/searcher_impl.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/pkg/search"
+	"github.com/stackrox/rox/pkg/search/blevesearch"
+	"github.com/stackrox/rox/pkg/search/paginated"
+	"github.com/stackrox/rox/pkg/search/sortfields"
+)
+
+var (
+	defaultSortOption = &v1.QuerySortOption{
+		Field:    search.DeploymentID.String(),
```
Comment (RTann) [inline central/imageintegration/datastore/internal/search/searcher_impl.go]: image integrations do not have a `DeploymentID`. Perhaps you can just do `ClusterID` instead

File: central/imageintegration/datastore/singleton.go [hunk 5]
```diff
@@ -3,8 +3,11 @@ package datastore
 import (
 	"context"
 
-	"github.com/stackrox/rox/central/enrichment"
 	"github.com/stackrox/rox/central/globaldb"
+	"github.com/stackrox/rox/central/globalindex"
+	"github.com/stackrox/rox/central/imageintegration/datastore/search"
+
```
Comment (RTann) [inline central/imageintegration/datastore/singleton.go:9]: nit: delete newline

File: central/imageintegration/datastore/search/searcher.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/search"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+// Searcher provides search functionality on existing pods
```
Comment (RTann) [inline central/imageintegration/datastore/search/searcher.go:17]: nit: existing image integrations

File: central/imageintegration/datastore/search/searcher.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+	indexer "github.com/stackrox/rox/central/imageintegration/index"
+	"github.com/stackrox/rox/central/imageintegration/store"
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/search"
+)
+
+var (
+	log = logging.LoggerForModule()
```
Comment (RTann) [inline central/imageintegration/datastore/search/searcher.go:14]: is this used?

File: central/enrichment/singleton.go [hunk 8]
```diff
@@ -9,26 +10,29 @@ import (
 	nodeCVEDataStore "github.com/stackrox/rox/central/cve/node/datastore"
 	"github.com/stackrox/rox/central/image/datastore"
 	"github.com/stackrox/rox/central/imageintegration"
+	imageintegrationDataStore "github.com/stackrox/rox/central/imageintegration/datastore"
```
Comment (RTann) [inline central/enrichment/singleton.go:13]: nit: `imageIntegrationDataStore`

File: central/cluster/datastore/datastore_impl.go [hunk 9]
```diff
@@ -152,6 +155,24 @@ func (ds *datastoreImpl) UpdateClusterStatus(ctx context.Context, id string, sta
 	return ds.clusterStorage.Upsert(ctx, cluster)
 }
 
+func (ds *datastoreImpl) removeImageIntegrationByCluster(ctx context.Context, cluster *storage.Cluster) {
```
Comment (RTann) [inline central/cluster/datastore/datastore_impl.go:158]: nit: Let's try to keep it consistent with the other related functions: `removeClusterImageIntegrations`

Comment (RTann) [reply central/cluster/datastore/datastore_impl.go:158]: Also maybe move this directly below `postRemoveCluster` with the rest of the related functions

File: central/cluster/datastore/datastore_impl.go [hunk 10]
```diff
@@ -516,6 +537,7 @@ func (ds *datastoreImpl) postRemoveCluster(ctx context.Context, cluster *storage
 		}
 	}
 
+	ds.removeImageIntegrationByCluster(ctx, cluster)
```
Comment (RTann) [inline central/cluster/datastore/datastore_impl.go:540]: nit: new line after this

File: central/cluster/datastore/datastore_impl_test.go [hunk 11]
```diff
@@ -127,8 +129,9 @@ func (suite *ClusterDataStoreTestSuite) SetupTest() {
 	suite.clusters.EXPECT().Walk(gomock.Any(), gomock.Any()).Return(nil)
 	suite.healthStatuses.EXPECT().Walk(gomock.Any(), gomock.Any()).Return(nil)
 	suite.indexer.EXPECT().AddClusters(nil).Return(nil)
-
+	suite.imageIntegrationDataStore = imageIntegrationDatastoreMocks.NewMockDataStore(suite.mockCtrl)
 	var err error
+
```
Comment (RTann) [inline central/cluster/datastore/datastore_impl_test.go:134]: nit: remove new line

File: central/cluster/datastore/datastore_impl_test.go [hunk 12]
```diff
@@ -127,8 +129,9 @@ func (suite *ClusterDataStoreTestSuite) SetupTest() {
 	suite.clusters.EXPECT().Walk(gomock.Any(), gomock.Any()).Return(nil)
 	suite.healthStatuses.EXPECT().Walk(gomock.Any(), gomock.Any()).Return(nil)
 	suite.indexer.EXPECT().AddClusters(nil).Return(nil)
-
+	suite.imageIntegrationDataStore = imageIntegrationDatastoreMocks.NewMockDataStore(suite.mockCtrl)
```
Comment (RTann) [inline central/cluster/datastore/datastore_impl_test.go:132]: nit: re-add new line below this. Also, this should be with the other related initializers above

File: central/cve/fetcher/manager_impl_postgres_test.go [hunk 13]
```diff
@@ -22,6 +22,7 @@ import (
 	mockCVEDataStore "github.com/stackrox/rox/central/cve/cluster/datastore/mocks"
 	clusterCVESearch "github.com/stackrox/rox/central/cve/cluster/datastore/search"
 	clusterCVEPostgres "github.com/stackrox/rox/central/cve/cluster/datastore/store/postgres"
+
```
Comment (RTann) [inline central/cve/fetcher/manager_impl_postgres_test.go:25]: why the new line?

File: central/enrichment/singleton.go [hunk 14]
```diff
@@ -51,8 +55,20 @@ func initialize() {
 		signatureIntegrationDataStore.Singleton().GetAllSignatureIntegrations)
 	ne = nodeEnricher.New(nodeCVESuppressor, metrics.CentralSubsystem)
 	en = New(datastore.Singleton(), ie)
+	ad = imageIntegrationDS.Singleton()
```
Comment (RTann) [inline central/enrichment/singleton.go:58]: nit: initialize this closer to where it's used (ie below where the `newManager` is made)

**Review by RTann — COMMENTED** [review]
Some comments for now. I see you are still testing, but wanted to let you know some initial thoughts. Let em know when you are ready for a full review

File: central/cluster/datastore/datastore.go [hunk 15]
```diff
@@ -9,6 +9,7 @@ import (
 	clusterStore "github.com/stackrox/rox/central/cluster/store/cluster"
 	clusterHealthStore "github.com/stackrox/rox/central/cluster/store/clusterhealth"
 	deploymentDataStore "github.com/stackrox/rox/central/deployment/datastore"
+	imageIntegrationStore "github.com/stackrox/rox/central/imageintegration/datastore"
```
Comment (md2119) [inline central/cluster/datastore/datastore.go:12]: nit: `imageIntegrationDataStore`

File: central/imageintegration/datastore/datastore_impl.go [hunk 16]
```diff
@@ -3,20 +3,41 @@ package datastore
...
-	storage store.Store
+	storage  store.Store
+	indexer  index.Indexer
+	searcher search.Searcher
+}
+
+func (ds *datastoreImpl) Search(ctx context.Context, q *v1.Query) ([]search.Result, error) {
+	integrations, err := ds.searcher.Search(ctx, q)
+	if err != nil {
+		log.Error(err.Error() + " Failed to search image integrations")
```
Comment (md2119) [inline central/imageintegration/datastore/datastore_impl.go:37]: `log.Errorf`

File: central/imageintegration/datastore/datastore_impl.go [hunk 17]
```diff
@@ -89,3 +110,96 @@ func (ds *datastoreImpl) RemoveImageIntegration(ctx context.Context, id string)
 
 	return ds.storage.Delete(ctx, id)
 }
+
+func (ds *datastoreImpl) fullReindex(ctx context.Context) error {
+	log.Info("[STARTUP] Reindexing all pods")
```
Comment (md2119) [inline central/imageintegration/datastore/datastore_impl.go:115]: If postgres is enabled, return

**Review by md2119 — COMMENTED** [review]
I am confused. The JIRA requests a fix for orphaned image integrations where as this PR also adds a postgres store. I recommend splitting up the work:
First, address the issue that JIRA describes
Second, add postgres store for image integrations

File: central/cluster/datastore/datastore_impl.go [hunk 18]
```diff
@@ -152,6 +155,24 @@ func (ds *datastoreImpl) UpdateClusterStatus(ctx context.Context, id string, sta
 	return ds.clusterStorage.Upsert(ctx, cluster)
 }
 
+func (ds *datastoreImpl) removeImageIntegrationByCluster(ctx context.Context, cluster *storage.Cluster) {
+
+	q := pkgSearch.NewQueryBuilder().AddExactMatches(pkgSearch.ClusterID, cluster.GetId()).ProtoQuery()
+	log.Info(">>>> Image Integrations: " + q.String() + " " + q.GetBaseQuery().String())
+	imageIntegrations, err := ds.imageIntegrationDataStore.Search(ctx, q)
+	log.Infof(">>>>Image Integrations list size is %d", len(imageIntegrations))
```
Comment (rhybrillou) [inline central/cluster/datastore/datastore_impl.go:163]: If the two log lines here are for local debugging purpose, make sure to remove them before merging.

File: central/imageintegration/datastore/datastore_impl.go [hunk 19]
```diff
@@ -89,3 +102,21 @@ func (ds *datastoreImpl) RemoveImageIntegration(ctx context.Context, id string)
 
 	return ds.storage.Delete(ctx, id)
```
Comment (rhybrillou) [inline central/imageintegration/datastore/datastore_impl.go:103]: You probably want to remove the item from the index and the storage here, and push the item to the index on Addition, Update or Upsert.
The pattern I saw in `alert` is
- For Add/Update/Upsert:
```
if err := ds.storage.CUfunction(ctx, object); err != nil { return err }
if err := ds.index.WriteToIndex(objectToIndex); err != nil { return err }
return ds.storage.AckKeysIndexed(ctx, object.GetId())
```
- For deletion:
```
errorList := errorhelpers.NewErrorList("deleting")
if err := ds.storage.Delete(ctx, objId); err != nil { errorList.AddError(err) }
if err := ds.indexer.Delete(objid); err != nil { errorList.AddError(err) }
if err := ds.storage.AckKeysIndexed(ctx, objid); err != nil { errorList.AddError(err) }
return errorList.ToError()
```

**Review by rhybrillou — COMMENTED** [review]
Please make sure the index is updated in the write operations so search can work.
I'll have a second look once indexing is fixed.

File: central/imageintegration/datastore/datastore_impl.go (hunk 19 above)
Comment (daynewlee) [reply central/imageintegration/datastore/datastore_impl.go:107]: I could be wrong. But AckKeysIndexed is not available for Bolt as far as I've seen.

**Pull request closed without merging**